### ISSUE TITLE
Import issue with auto schema creation with dataclass

### ISSUE THREAD

**Issue by lpgros** [pydantic/pydantic#8475]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Hi,

I boiled down 2 issues (that look like just 1 really) to these 2 files and steps.

When running file 1 as is, I get the following error:
```
pydantic.errors.PydanticUndefinedAnnotation: name 'Optional' is not defined
```

if I import manually `Optional` in file 1, then it "works".
In a seemingly, related issue, at this point, I get this error:
```
pydantic.errors.PydanticUndefinedAnnotation: name 'datetime' is not defined
```
which can be corrected by changing the import to 
```Python
import datetime
```
and updating the usages to
```Python
Optional[datetime.datetime]
```

All in all, it looks like Pydantic does not use/execute the imports for file 2 when running file 1 even though it uses file 2

Having 2 files is required for the issue to reproduce (all in 1 file does not crash).

### Example Code

file 1:

```Python
from __future__ import annotations


from my_class import MyClass
from pydantic import ConfigDict, validate_call


@validate_call(
    config=ConfigDict(strict=True, arbitrary_types_allowed=True),
    validate_return=True,
)
def test(my_class: MyClass):
    pass
```

file 2:
```Python
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Optional


@dataclass
class MyClass:
    member: Optional[str] = None
    created: Optional[datetime] = None
    updated: Optional[datetime] = None
    deleted: Optional[datetime] = None
```

### Python, Pydantic & OS Version

```Text
$ python -c "import pydantic.version; print(pydantic.version.version_info())"

             pydantic version: 2.5.3
        pydantic-core version: 2.14.6
          pydantic-core build: profile=release pgo=false
                 install path: .virtualenvs/env3/lib/python3.9/site-packages/pydantic
               python version: 3.9.12 (main, May 12 2023, 09:21:14)  [Clang 14.0.3 (clang-1403.0.22.14.1)]
                     platform: macOS-14.1.2-x86_64-i386-64bit
             related packages: email-validator-2.0.0.post2 typing_extensions-4.7.1
```

**Comment by sydney-runkle** [comment]
@lpgros,

If you remove `from __future__ import annotations` from `file2`, things work as expected. This doesn't seem to be an issue with `pydantic`, but rather your import patterns.

**Comment by lpgros** [comment]
@sydney-runkle `annotations` is needed when using forward references with type hinting and an integral part of the Python standard library.

It is listed as "working" in the Pydantic doc: https://docs.pydantic.dev/latest/concepts/postponed_annotations/

How could I have a fully typehinted project without them?

Please advise

**Comment by sydney-runkle** [comment]
@lpgros,

Taking a closer look at this, I think there is something we can do on our end to fix this issue. We'll open a PR shortly. Thanks for following up!

**Comment by lpgros** [comment]
cheers! That would be great 😄

**Comment by sydney-runkle** [comment]
@lpgros,

Sure thing. When investigating more, we discovered that this is related to how we deal with namespaces for dataclasses. We have more intentional handling of namespaces for `BaseModel` types, so it's not surprising that this issue popped up for dataclasses. I suspect it might be an issue for `TypedDict` types as well.

Additionally, this problem crops up because of limitations with processing annotations that have been resolved with later Python versions. Stay tuned for our PR :).
